OpenTrons/opentrons-api
# feat(api): Expose firmware update progress to the API layer

# Overview

We need to expose firmware update progress to the API layer, this pull request will accomplish that.
Closes: [RCORE-480](https://opentrons.atlassian.net/browse/RCORE-480)

# Test Plan
- Make sure updates still work as expected
- Make sure firmware update progress is reported when calling OT3controller.firmware_update
- Make sure firmware updates are reported when calling api.get_fw_update_progress

# Changelog

- Controller.update_firmware now returns AsyncIterator containing the UpdateStatus and total progress
- added _received_update_progress which parses and converts RunUpdate.run_updates output into new UpdateStatus and update_progress
- _received_update_progress caches UpdateStatus and update_progress to keep track of updates
- added get_update_progress to ot3Controller to get the UpdateStatus and update_progress cached by _received_update_progress

# Review requests

- does this make sense so far?
- is there anything I'm overlooking?

# Risk assessment
Low, OT3 dev work

# Todo
- [ ] maybe rebase to get rid of all these unrelated, already-in-edge commits?

[RCORE-480]: https://opentrons.atlassian.net/browse/RCORE-480?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
